homeassistant/components/mqtt/mixins.py

```diff
@@ -513,12 +517,170 @@ async def cleanup_device_registry(hass, device_id, config_entry_id):
         and not tag.async_has_tags(hass, device_id)
     ):
         device_registry.async_update_device(
-            device_id, remove_config_entry_id=config_entry_id
+            device_id, remove_config_entry_id=cast(str, config_entry_id)
         )
 
 
+def get_discovery_hash(discovery_data: dict | None) -> tuple:
+    """Get the discovery hash from the discovery data."""
+    return discovery_data[ATTR_DISCOVERY_HASH] if discovery_data else ()
+
+
+def send_discovery_done(hass: HomeAssistant, discovery_data: dict | None) -> None:
+    """Acknowledge a discovery message has been handled."""
+    if discovery_hash := get_discovery_hash(discovery_data):
+        async_dispatcher_send(hass, MQTT_DISCOVERY_DONE.format(discovery_hash), None)
+
+
+def terminate_discovery(
+    hass: HomeAssistant, discovery_data: dict | None, dispatcher: Callable | None
+) -> None:
+    """Terminate the discovery."""
+    if (discovery_hash := get_discovery_hash(discovery_data)) == ():
+        return
+    if dispatcher:
+        dispatcher()
+        dispatcher = None
+    clear_discovery_hash(hass, discovery_hash)
+
+
+async def async_remove_discovery_payload(hass: HomeAssistant, discovery_data: dict):
+    """Remove the auto discovery payload to avoid rediscovery after a restart."""
+    discovery_topic = discovery_data[ATTR_DISCOVERY_TOPIC]
+    await async_publish(hass, discovery_topic, "", retain=True)
+
+
+class MqttDiscoveryDeviceUpdate:
+    """Add support for auto discovery for platforms without an entity."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        discovery_data: dict,
+        device_id: str,
+        config_entry: ConfigEntry,
+        log_name: str,
+    ) -> None:
+        """Initialize the update service."""
+
+        # Only activate if the parent class has a discovery hash
+        self.hass = hass
+        self.log_name = log_name
+        self._remove_discovery = None
+
+        self._discovery_data = discovery_data
+
+        config_entry_id = config_entry.entry_id
+        rediscover: bool = False
+        skip_device_removal: bool = False
+
+        async def async_discovery_update(
+            discovery_payload: DiscoveryInfoType | None,
+        ) -> None:
+            """Handle discovery update."""
+            nonlocal rediscover, self
+            discovery_hash = get_discovery_hash(discovery_data)
+            _LOGGER.info(
+                "Got update for %s with hash: %s '%s'",
+                self.log_name,
+                discovery_hash,
+                discovery_payload,
+            )
+            if (
+                discovery_payload
+                and discovery_payload != self._discovery_data[ATTR_DISCOVERY_PAYLOAD]
+            ):
+                rediscover = True
+            if not discovery_payload or rediscover:
+                # unregister and clean up the current discovery instance
```

Comment:
Both trigger and tag had support for being updated, looks like that's gone now?
I'm not sure this is wanted, it means we'll tear everything down and set it up again if there's a name change for example, it means we introduce a "subscribe hole" where we'll miss MQTT messages sent for the tag or trigger.
What's the rationale for dropping the update support?